File: DesignPatterns/01_GOF23/DP_examples/Go/01_Creational/Factory/Creator.go
package Factory

// ICreator 创建者接口
type ICreator interface {
	GetShape() IShape
}

// unknownShapeCreator 未知形状创建者
type unknownShapeCreator struct{}

func (c unknownShapeCreator) GetShape() IShape {
	return &UnknownShape{}
}

// circleCreator 圆形创建者
type circleCreator struct{}

func (c circleCreator) GetShape() IShape {
	return &Circle{}
}

// squareCreator 正方形创建者
type squareCreator struct{}

func (c squareCreator) GetShape() IShape {
	return &Square{}
}

// rectangleCreator 矩形创建者
type rectangleCreator struct{}

func (c rectangleCreator) GetShape() IShape {
	return &Rectangle{}
}

// bigCircleCreator 大圆形创建者
type bigCircleCreator struct {
	circleCreator
}

func (c bigCircleCreator) setRadius() {
	println("Set Radius to 10")
}

func (c bigCircleCreator) GetShape() IShape {
	c.setRadius()
	return c.circleCreator.GetShape()
}

func newCreator(kind ShapeKind) ICreator {
	var creator ICreator
	switch kind {
	case CIRCLE:
		creator = circleCreator{}
	case SQUARE:
		creator = squareCreator{}
	case RECTANGLE:
		creator = rectangleCreator{}
	case BIG_CIRCLE:
		creator = bigCircleCreator{}
	default:
		creator = unknownShapeCreator{}
	}
	return creator
}

File: DesignPatterns/01_GOF23/DP_examples/CSharp/01_Creational/Factory/Creator.cs
namespace Creational.Factory;

public interface ICreator {
    IShape getShape();
    static virtual ICreator Instance { get; }
}

public sealed partial class ShapeFactory {
    class Creator {
        public static ICreator GetCreator(ShapeKind kind) {
            return kind switch {
                ShapeKind.Circle => CircleCreator.Instance,
                ShapeKind.BigCircle => BigCircleCreator.Instance,
                ShapeKind.Rectangle => RectangleCreator.Instance,
                ShapeKind.Square => SquareCreator.Instance, 
                _ => UnknownShapeCreator.Instance,
            };
        }
    }
    internal class UnknownShapeCreator : ICreator {
        public static ICreator Instance { get; } = new UnknownShapeCreator();
        public virtual IShape getShape() => new UnknownShape();
    }
    internal class CircleCreator : ICreator {
        public static ICreator Instance { get; } = new CircleCreator();
        public virtual IShape getShape() => new Circle();
    }
    internal class SquareCreator : ICreator {
        public static ICreator Instance { get; } = new SquareCreator();
        public virtual IShape getShape() => new Square();
    }
    internal class RectangleCreator : ICreator {
        public static ICreator Instance { get; } = new RectangleCreator();
        public virtual IShape getShape() => new Rectangle();
    }
    // 扩展 Circle
    internal class BigCircleCreator : CircleCreator, ICreator {
        public static new ICreator Instance => new BigCircleCreator();
        private IShape SetRadius(IShape circle) {
            // use circle
            Console.WriteLine("Set Radius to 10");
            return circle;
        }
        public override IShape getShape() => SetRadius(base.getShape());
    }

   
}
